src/Behat/Testwork/Exception/Stringer/PHPUnitExceptionStringer.php

```diff
@@ -35,17 +35,24 @@ public function supportsException(Exception $exception)
      */
     public function stringException(Exception $exception, $verbosity)
     {
-        if (class_exists('PHPUnit\\Util\\ThrowableToStringMapper')) {
-            return trim(\PHPUnit\Util\ThrowableToStringMapper::map($exception));
-        }
-
-        if (!class_exists('PHPUnit\\Framework\\TestFailure')) {
-            return trim(\PHPUnit_Framework_TestFailure::exceptionToString($exception));
-        }
-
         // PHPUnit assertion exceptions do not include expected / observed info in their
         // messages, but expect the test listeners to format that info like the following
         // (see e.g. PHPUnit_TextUI_ResultPrinter::printDefectTrace)
-        return trim(\PHPUnit\Framework\TestFailure::exceptionToString($exception));
+
+        switch (true) {
+            case class_exists(\PHPUnit\Util\ThrowableToStringMapper::class):
+                return trim(\PHPUnit\Util\ThrowableToStringMapper::map($exception));
+
+            case class_exists(\PHPUnit\Framework\TestFailure::class):
+                return trim(\PHPUnit\Framework\TestFailure::exceptionToString($exception));
```

Comment:
These classes are both [annotated that they're not covered by the phpunit Backward Compatibility promise](https://github.com/sebastianbergmann/phpunit/blob/f72a244153efe365f97c053f7b328ec811b93fe9/src/Util/ThrowableToStringMapper.php#L22C4-L22C89).

I don't think there's an easy way to replicate the phpunit behaviour while guaranteeing BC. But (as with my other comment) I think we should do our best to return some sort of string that reflects the user's own test failure even if e.g. they upgrade to a new phpunit with a different method signature. So we should wrap this whole thing in a `try {...} catch (Throwable $e) { return 'a useful string incorporating the original exception class & message'}`.

Reply:
If we want to be absolutely safe (by catching all exceptions), I would still consider rethrowing an exception with the original exception as "previous" (as per my previous comment). That should provide the user with all the relevant info and feels like a rather standard way to do things.